internal/types/path.go

```diff
@@ -17,17 +17,87 @@ package types
 import (
 	"fmt"
 	"strconv"
+	"strings"
 )
 
+// Path represents field path type. It should be used wherever we work with paths or dot notation.
+// Path should be stored and passed as value. All of its methods return new values, not modifying the receiver's state.
+type Path struct {
+	s []string
+}
+
+// NewPath returns Path from a strings slice.
+func NewPath(path []string) Path {
+	p := Path{s: make([]string, len(path))}
+	copy(p.s, path)
+	return p
+}
+
+// NewPathFromString returns Path from path string. Path string should contain fields separated with '.'.
+func NewPathFromString(s string) Path {
+	path := strings.Split(s, ".")
+
+	return NewPath(path)
+}
+
+// String returns dot-separated path value.
+func (p Path) String() string {
+	return strings.Join(p.s, ".")
+}
+
+// Len returns path length.
+func (p Path) Len() int {
+	return len(p.s)
+}
+
+// Slice returns path values array.
+func (p Path) Slice() []string {
+	path := make([]string, p.Len())
+	copy(path, p.s)
+	return path
+}
+
+// Suffix returns the last path element.
+func (p Path) Suffix() string {
+	if len(p.s) == 0 {
+		return ""
+	}
+	return p.s[p.Len()-1]
+}
+
+// Prefix returns the first path element.
+func (p Path) Prefix() string {
+	if p.Len() == 0 {
+		return ""
+	}
+	return p.s[0]
+}
+
+// TrimSuffix returns a path without the last element.
+func (p Path) TrimSuffix() Path {
+	if p.Len() == 0 {
+		return p
```

Comment:
Here is also misleading. I expect if there is nothing to trim I must not trim.

Reply:
That's exactly what is happening here. I there is no `Path` elements we not trim anything and return the path itself. It's the caller responsibility to track if there is something in the path left to trim not the callee. The only option might be here IMO is to return a new path calling `NewPath([]string{})`

Reply:
To the function name, yes.
But looking at the usages in this PR - no, you suppose there must have been an element.